core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -2952,6 +2961,112 @@ class KafkaController(val config: KafkaConfig,
       allocateProducerIdsRequest.brokerEpoch, eventManagerCallback))
   }
 
+  def processCorruptedBrokersChange(): Unit = {
+    if (!isActive) {
+      return
+    }
+    val corruptedBrokers = zkClient.getCorruptedBrokers.mapValues(_.clearedFromIsrs)
+    info(s"Received corrupted brokers change = $corruptedBrokers")
+    if (corruptedBrokers == controllerContext.corruptedBrokers) {
+      return
+    }
+    maybeCleanIsrsForCorruptedBrokers(corruptedBrokers)
+  }
+
+  private def maybeCleanIsrsForCorruptedBrokers(originalCorruptedBrokers: Map[Int, Boolean]): Unit = {
+    // If delayed election is not enabled, do not clear corrupted broker from ISR. If it is removed, it
+    // may cause problems with elections when unclean leader election is disabled.
+    if (!config.isDelayedElectionEnabled) {
+      return
+    }
+
+    def updateCorruptedBrokerToCleanedInZk(brokerId: Int): Try[Unit] = Try {
+      info(s"Updating zk node for corrupted broker $brokerId to cleaned")
+      val corruptedBroker = CorruptedBroker(brokerId, clearedFromIsrs = true)
+      zkClient.updateCorruptedBroker(corruptedBroker)
+    }
+
+    val corruptedBrokersToClearFromIsr = originalCorruptedBrokers
+      .filterNot { case (_, isrCleared) => isrCleared }.keys
+    info(s"Clearing corrupted brokers $corruptedBrokersToClearFromIsr from ISR")
+    val corruptedBrokersIsrClearResults = corruptedBrokersToClearFromIsr.map { corruptedBrokerId =>
+      val tryResult = removeBrokerIdFromPartitionsIsrs(corruptedBrokerId).flatMap { _ =>
+        updateCorruptedBrokerToCleanedInZk(corruptedBrokerId)
+      }
+      corruptedBrokerId -> tryResult
+    }
+
+    val (isrCleanSuccessful, isrCleanUnsuccessful) = corruptedBrokersIsrClearResults.partition(_._2.isSuccess)
+    error(s"Could not clear corrupted brokers $isrCleanUnsuccessful from ISR")
```

Comment:
Shouldn't this be conditional on a non-empty isrCleanUnsuccessful?

Reply:
Oh right! My bad.

Reply:
Fixed.